src/ZODB/mvccadapter.py

```diff
@@ -151,8 +151,19 @@ def poll_invalidations(self):
 
     def load(self, oid):
         assert self._start is not None
-        r = self._storage.loadBefore(oid, self._start)
-        if r is None:
+        data, serial = loadBeforeEx(self._storage, oid, self._start)
+        if data is None:
+            # raise POSKeyError if object does not exist at all
+            # TODO raise POSKeyError always and switch to raising ReadOnlyError only when
+            # actually detecting that load is being affected by simultaneous pack (see below).
+            if serial == z64:
+                # XXX second call to loadBeforeEx - it will become unneeded once we
+                # switch to raising POSKeyError.
+                _, serial_exists = loadBeforeEx(self._storage, oid, maxtid)
+                if serial_exists == z64:
+                    # object does not exist at all
+                    raise POSException.POSKeyError(oid)
+
             # object was deleted or not-yet-created.
             # raise ReadConflictError - not - POSKeyError due to backward
```

Comment:
Thanks for asking.

This is indeed a bit hairy place, and I would be more happier if we could just
raise `POSKeyError` unconditionally here. We _have_ to raise `POSKeyError` when
load finds that there is no such object, because this is logically expected
behaviour when application triggers load of "not existing" object. Previously,
for "object does not exist at all", this behaviour was implemented by
`loadBefore` itself:

https://github.com/zopefoundation/ZODB/blob/1fb097b41cae4ca863c8a3664414c9ec0e204393/src/ZODB/interfaces.py#L710-L724

However raising just `POSKeyError`, when load finds that an object was deleted
as of `MVCCAdapterInstance`'s view of the database, leads to the following test
failures:

```diff
--- a/src/ZODB/mvccadapter.py
+++ b/src/ZODB/mvccadapter.py
@@ -153,6 +153,7 @@ def load(self, oid):
         assert self._start is not None
         data, serial = loadBeforeEx(self._storage, oid, self._start)
         if data is None:
+            raise POSException.POSKeyError(oid)
             # raise POSKeyError if object does not exist at all
             # TODO raise POSKeyError always and switch to raising ReadOnlyError only when
             # actually detecting that load is being affected by simultaneous pack (see below).
```

(1):
```
File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/testmvcc.py", line 413, in ZODB.tests.testmvcc
Failed example:
    r1["b"]._p_activate() # doctest: +ELLIPSIS
Expected:
    Traceback (most recent call last):
     ...
    ReadConflictError: ...
Got:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest ZODB.tests.testmvcc[137]>", line 1, in <module>
        r1["b"]._p_activate() # doctest: +ELLIPSIS
      File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/Connection.py", line 777, in setstate
        p, serial = self._storage.load(oid)
      File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/mvccadapter.py", line 156, in load
        raise POSException.POSKeyError(oid)
    POSKeyError: 0x02
```

(2):
```
Error in test checkPackNowWhileWriting (ZODB.tests.testMappingStorage.MappingStorageTests)
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/util.py", line 265, in wrapper
    return func(*args, **kwargs)
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/PackableStorage.py", line 284, in checkPackNowWhileWriting
    self._PackWhileWriting(pack_now=True)
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/PackableStorage.py", line 229, in _PackWhileWriting
    t.join(60)
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/MTStorage.py", line 45, in join
    self._exc_info[0], self._exc_info[1], self._exc_info[2])
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/MTStorage.py", line 37, in run
    self.runtest()
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/PackableStorage.py", line 764, in runtest
    conn.root()[index].value = MinPO(j)
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/Connection.py", line 777, in setstate
    p, serial = self._storage.load(oid)
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/mvccadapter.py", line 156, in load
    raise POSException.POSKeyError(oid)
POSKeyError: 0x0a
```

The first test explicitly verifies that `ReadConflictError` is raised when
`Connection` is requested to retrieve object that exists, but was not yet
created as of connection's view of the database. It was added in 2004 in
https://github.com/zopefoundation/ZODB/commit/020daaf4cee8.

The second failure in [checkPackNowWhileWriting](https://github.com/zopefoundation/ZODB/blob/1fb097b41cae/src/ZODB/tests/PackableStorage.py#L183-L284) - that happens only sometimes - is more illustrative to "concurrent pack" problem:

- one long-running transaction is accessing the database via ZODB connection as of `@txn1` state.
- during that time, via other connection, the database is mutated with a new committed transactions (txn2) that changes object, that was already present in the database.
- then `pack(t=txn2)` is run. The pack decides to keep `obj@txn2` and deletes `obj@txn1` as instructed.
- the first transaction is still running, tries to access obj.
- load obj@txn1 finds "object exists, but was not yet created".

In other words, due to the way MVCC is implemented (snapshotted views of the
database), and in the presence of simultaneous changes, a Connection is usually
observing the database close, but not exactly, to its real head - lagging
behind it a bit. Then if pack is ran with some `t` that is inside [`conn.at`, `lastTransaction`],
for that connection it looks like the pack - concurrently to read activities of
the connection - deleted some objects under the foot. It is kind of
reader/writer conflict error.

As explained in https://github.com/zopefoundation/ZODB/commit/3a493b010876
(both commit message and changes in `mvccadapter.py`), such situations are
reported to outside via raising `ReadConflictError`.

In applications servers (e.g. Zope), transactions that raise `ConflictError`
are automatically restarted:

https://github.com/zopefoundation/ZODB/blob/1fb097b41cae4ca863c8a3664414c9ec0e204393/src/ZODB/transact.py#L25-L62

This way, if concurrent pack deletes an object under the foot of a connection,
due to `ReadConflictError` being raised, it will lead to the transaction to be
silently restarted without any error reported to user.

However, if we change to stop raising that `ReadConflictError` and instead
raise `POSKeyError` unconditionally, a real error - that `POSKeyError` - will
be reported to the user and to the log.

That's why, for backward-compatibility, the new code goes in the direction to
preserve previous behaviour of rasing `ReadConflictError` for the case when
object "exists but was not yet created".

I hope it clarifies the story of "pack and why we have to peek into the future" a bit.

Please also see the following links for more details:

- discussion starting from https://github.com/zopefoundation/ZODB/pull/320#issuecomment-661819221,
- discussion in https://github.com/zopefoundation/ZODB/pull/322.

Reply:
Satisfied with your explanation. I did not get it from the existing source code comment.

The existence of the mvcc test indicates that it may have been a design goal to raise `ReadConflictError` in the case "object only after" -- independent of pack. Maybe, the motivation was that a restarted transaction would see the object.